src/PSR7/Validators/BodyValidator/BodyValidator.php

```diff
@@ -42,38 +49,50 @@ public function validate(OperationAddress $addr, MessageInterface $message) : vo
-
-        // As per https://tools.ietf.org/html/rfc7231#section-3.1.1.5 and https://tools.ietf.org/html/rfc7231#section-3.1.1.1
-        // ContentType can contain multiple statements (type/subtype + parameters), ie: 'multipart/form-data; charset=utf-8; boundary=__X_PAW_BOUNDARY__'
-        // OpenAPI Spec only defines the first part of the header value (type/subtype)
-        // Other parameters SHOULD be skipped
-        $contentType = strtok($contentType, ';');
 
         // does the response contain one of described media types?
         if (! isset($mediaTypeSpecs[$contentType])) {
             throw InvalidHeaders::becauseContentTypeIsNotExpected($contentType, $addr);
         }
 
-        // ok looks good, now apply validation
-        $body = (string) $message->getBody();
+        // detect the schema for the media type
+        $schema = $mediaTypeSpecs[$contentType]->schema;
+        if (! $schema) {
+            // no schema means no validation
+            // note: schema is REQUIRED to define the input parameters to the operation when using multipart content
+            return;
+        }
 
-        if (preg_match('#^application/.*json$#', $contentType)) {
-            $body = json_decode($body, true);
-            if (json_last_error() !== JSON_ERROR_NONE) {
-                throw InvalidBody::becauseBodyIsNotValidJson(json_last_error_msg(), $addr);
-            }
+        // Validate message body
+        if (preg_match('#^multipart/.*#', $contentType)) {
+            $this->validateMultipart($addr, $message, $mediaTypeSpecs, $contentType);
+        } elseif (preg_match('#^application/x-www-form-urlencoded$#', $contentType)) {
+            $this->validateFormUrlencoded($addr, $message, $mediaTypeSpecs, $contentType);
+        } else {
+            $this->validateUnipart($addr, $message, $mediaTypeSpecs, $contentType);
         }
+    }
 
-        $validator = new SchemaValidator($this->detectValidationStrategy($message));
-        try {
-            $validator->validate($body, $mediaTypeSpecs[$contentType]->schema);
-        } catch (SchemaMismatch $e) {
-            throw InvalidBody::becauseBodyDoesNotMatchSchema($contentType, $addr, $e);
+    private function messageContentType(MessageInterface $message) : ?string
+    {
+        $contentTypes = $message->getHeader(self::HEADER_CONTENT_TYPE);
+        if (! $contentTypes) {
+            return null;
         }
+
+        $contentType = $contentTypes[0]; // use the first value
+
+        // As per https://tools.ietf.org/html/rfc7231#section-3.1.1.5 and https://tools.ietf.org/html/rfc7231#section-3.1.1.1
+        // ContentType can contain multiple statements (type/subtype + parameters), ie: 'multipart/form-data; charset=utf-8; boundary=__X_PAW_BOUNDARY__'
+        // OpenAPI Spec only defines the first part of the header value (type/subtype)
+        // Other parameters SHOULD be skipped
+        $contentType = strtok($contentType, ';');
```

Comment:
This library grows a lot with code which purpose it `how to read PSR-7 message properly`. I'm not telling that the code is bad, but what if somewhere on internet there is a library for that particular responsibility? This would simplify things

Reply:
I am asking the same question myself.